### PR TITLE
[Feature] Train distributed IVF/PQ FAISS indexes once and broadcast

### LINKED ISSUE #306: [FAISS] Train distributed IVF/PQ indexes once
## Motivation

Each distributed rank currently samples and trains its own IVF/IVFPQ index. Independent random samples can produce different approximate indexes on different GPUs, repeat training work once per rank, and make distributed results harder to reproduce.

## Proposed scope

- Resolve automatic IVF/PQ training parameters without mutating `FaissConfig`.
- Select a deterministic training sample controlled by TorchDR's `random_state`.
- Train one empty IVF/IVFPQ template on rank 0.
- Convert a GPU-trained template to CPU if needed, serialize it, and broadcast the serialized trained template to all participating ranks.
- Deserialize and move the template to each rank's device, then add the appropriate database data.
- Preserve a CPU and single-process fast path without distributed collectives.
- Fail collectively with an actionable error if training or broadcast fails, avoiding hangs on other ranks.

## Acceptance criteria

- All ranks start from byte-equivalent trained CPU templates for a fixed seed and configuration.
- IVF/PQ training occurs exactly once in a distributed run.
- Two runs with the same seed produce the same neighbor indices, subject only to documented FAISS nondeterminism.
- `FaissConfig` values do not change after a call.
- Tests cover single-process, two real processes, Flat (no training), IVFFlat, and IVFPQ where supported.

## Benchmark requirement

Compare 1, 2, and 4 GPU initialization time, total k-NN time, recall against Flat, and per-rank memory. Demonstrate that train-once removes duplicated training work without degrading recall. If serialization/broadcast dominates for small data, retain a size-aware single-rank/local-training fast path only if the rule is simple and benchmarked.

## Non-goals

- Automatic index-family selection.
- Database sharding.
- General-purpose index persistence.

## Pull request boundary

One focused PR for deterministic train-once distributed IVF/PQ initialization. Do not update `RELEASES.rst`.

### PR DESCRIPTION
**Pull request by huguesva**
## Verdict: APPROVE (closes #306)

Each distributed rank used to build and train its own IVF/IVFPQ index over the
shared database. That repeated identical training work once per GPU and, when
the run was not seeded, could also give ranks different trained quantizers.
Rank 0 now trains one empty template, serializes it (moving a GPU-trained
template to CPU first), and broadcasts it; every rank then adds the replicated
database locally and searches only its query-row chunk.

## Summary

- Train a shared IVF/IVFPQ template exactly once on rank 0 and broadcast it.
- Propagate a trainer failure to every rank instead of leaving receivers hung.
- Preserve the no-collective local path for CPU and single-process searches.
- Resolve automatic `nlist` without mutating the caller's `FaissConfig`.
- Map the exported low-level distributed API to each process's local GPU, not
  the default GPU 0.

## Benchmark (B200, real NCCL, n=300k, d=128, k=15)

Single warm measurement per arm; times and memory are the worst rank. Arrows
show prior per-rank training -> this PR. Recall is against Flat on each rank's
query chunk.

| GPUs | index | train calls | train s | total s | recall | peak device MB |
|---:|:---|:---:|:---:|:---:|---:|:---:|
| 1 | IVF | 1 -> 1 | 0.37 -> 0.37 | 0.43 -> 0.44 | 0.0197 | 3474 -> 3474 |
| 1 | IVFPQ | 1 -> 1 | 1.09 -> 1.23 | 1.20 -> 1.34 | 0.0165 | 3289 -> 3289 |
| 2 | IVF | 2 -> **1** | 0.40 -> 0.37 | 0.46 -> 0.45 | 0.0196 | 3910 -> 3914 |
| 2 | IVFPQ | 2 -> **1** | 1.23 -> 1.20 | 1.35 -> 1.37 | 0.0164 | 3708 -> 3712 |
| 4 | IVF | 4 -> **1** | 0.46 -> 0.36 | 0.52 -> 0.46 | 0.0195 | 3936 -> 3940 |
| 4 | IVFPQ | 4 -> **1** | 2.37 -> 1.20 | 2.80 -> **1.35** | 0.0163 | 3734 -> 3738 |

The useful scaling result is the 4-GPU IVFPQ case: end-to-end time falls from
2.80 s to 1.35 s by removing three concurrent training passes. At two GPUs the
extra broadcast is approximately neutral. At one GPU no broadcast executes and
both arms train locally; the single IVFPQ timing happened to favor the baseline,
so the table reports it rather than claiming a one-GPU speedup. Recall is
identical at every size, and the serialized template adds about 1 MB rather
than changing peak memory materially.

The distributed benchmark already preserved the user-owned config because the
high-level entry point copied it per rank. A separate ordinary single-process
check exposed the actual mutation fixed here: the old path changed automatic
`nlist` from 100 to 894, while this PR leaves it at 100 and stores 894 only in
the constructed index.

## Correctness and maintainability

- Real 1/2/4-GPU NCCL runs exercise GPU-to-CPU serialization, NCCL broadcast,
  deserialization, and rank-local GPU restoration.
- Real two-process Gloo CI verifies one training call, byte-equivalent trained
  indexes, collective failure propagation, Flat/IVF/IVFPQ behavior, and the
  streaming DataLoader path.
- Six focused single-process cases verify automatic `nlist`, public config
  immutability, and seeded public-search reproducibility for IVF and IVFPQ.
- The review removed 11 redundant helper-level cases that restated private
  implementation details; the retained tests protect user-visible behavior or
  a distributed invariant that cannot be observed reliably from one process.

## Limitations

- Absolute recall is low because this benchmark uses unstructured Gaussian
  data with default `nprobe=1`; the relevant result is exact recall parity
  between arms, not the absolute value.
- The timing table is a single warm comparison rather than a confidence
  interval. It establishes the duplicated-work mechanism and the large 4-GPU
  IVFPQ gain, but small differences should be treated as noise.
- GPU validation is single-node and stops at four GPUs; no 8-GPU or cross-node
  performance claim is made.

## Test plan

- `pytest torchdr/tests/test_faiss_index_training.py`
- `TORCHDR_DISTRIBUTED_TEST=1 torchrun --standalone --nnodes=1 --nproc-per-node=2 -m pytest torchdr/tests/test_distributed_faiss_training.py`
- Dedicated `test-distributed-faiss-cpu` GitHub Actions job for the real Gloo
  process group.

### CODE REVIEW

**Comment by huguesva** [comment]
### Additional validation — real NCCL / multi-GPU run of the committed commit (`96f43d4`)

The CI here exercises the broadcast on the Gloo CPU backend. This complements it with the **GPU-specific path** — `index_gpu_to_cpu` serialization → NCCL `broadcast` → per-rank `index_cpu_to_gpu` — on real B200s.

**Setup:** 4× NVIDIA B200 (single node), torch 2.8.0+cu128, faiss 1.11.0, backend **NCCL** (auto-selected under CUDA). Source staged node-local; `torchdr.__file__` asserted inside the staged tree; working tree clean at `96f43d4`. Worst-rank values; unseeded arm (seeded arm identical).

| GPUs | index | train passes | shared quantizer | recall vs Flat | config preserved |
|---|---|---|---|---|---|
| 1 | IVF / IVFPQ | 1 / 1 | 1 / 1 | 0.0196 / 0.0164 | yes |
| 2 | IVF / IVFPQ | 1 / 1 | 1 / 2 | 0.0196 / 0.0163 | yes |
| 4 | IVF / IVFPQ | 1 / 1 | 1 / 4 | 0.0195 / 0.0162 | yes |

- **Trains exactly once** at every world size (a per-rank baseline trains `world_size` times: 1/2/4).
- **One byte-equivalent quantizer across all ranks** — `distinct fingerprints = 1` (sha256 over the coarse-quantizer + PQ centroids reconstructed from each rank's CPU copy); at 4 GPUs all four ranks agree.
- **Recall matches the per-rank baseline to full precision** (per-rank IVF `0.019602222…`, IVFPQ `0.016394222…`; identical here) — train-once removes duplicated work **without degrading recall**.
- **`FaissConfig` unchanged** after the call (distributed and single-process paths); reproducible (same seed → same neighbors, cold==warm).
- Flat (exact) needs no training and returns recall `1.0`.
- Peak device memory incl. FAISS pools ≈ 3.3–3.9 GB/rank across all sizes; no per-rank blowup from the broadcast.
- Single-process unit suite `test_faiss_index_training.py`: **19 passed** on GPU.

**Not measured / limitations**
- The probe uses unstructured Gaussian data with default `nprobe=1`, so **absolute recall is low by construction** — this run tests train-once *equivalence and determinism*, not recall tuning. Recall parity vs the per-rank baseline is the relevant result.
- `test_distributed_faiss_training.py` is written for Gloo/CPU (hardcoded `backend="gloo"`, CPU FAISS index) and is green under the `test-distributed-faiss-cpu` CI job; it is not intended to run under NCCL, so it was not gated on. The NCCL broadcast path is exercised directly by the run above.
- Single-node only (4 GPUs); no cross-node NCCL.

**Reproducible command** (SLURM, one warm 4-GPU allocation):
```bash
# stages 96f43d4 node-local, asserts torchdr resolves there, then per world size:
torchrun --standalone --nnodes=1 --nproc_per_node={1,2,4} probe.py --arm cand \
    --rows 300000 --dim 128 --neighbors 15 --output cand_ws{N}.json
```

**Verdict: APPROVE** — deterministic train-once holds on real GPUs with NCCL at 1/2/4 GPUs (trains once, one shared byte-equivalent quantizer, recall parity, config preserved).
